src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,190 @@
 //! Basic HUGR quantum operations
 
+use std::cmp::max;
+use std::f64::consts::TAU;
+use std::num::NonZeroU64;
+
 use smol_str::SmolStr;
 
-use crate::extension::prelude::{BOOL_T, QB_T};
+use crate::extension::prelude::{BOOL_T, ERROR_TYPE, QB_T};
 use crate::extension::{ExtensionId, SignatureError};
 use crate::std_extensions::arithmetic::float_types::FLOAT64_TYPE;
 use crate::type_row;
-use crate::types::type_param::TypeArg;
-use crate::types::FunctionType;
+use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+use crate::types::{ConstTypeError, CustomCheckFailure, CustomType, FunctionType, Type, TypeBound};
+use crate::utils::collect_array;
+use crate::values::CustomConst;
 use crate::Extension;
 
 use lazy_static::lazy_static;
 
 /// The extension identifier.
 pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("quantum");
+
+/// Identifier for the angle type.
+const ANGLE_TYPE_ID: SmolStr = SmolStr::new_inline("angle");
+
+fn angle_custom_type(log_denom_arg: TypeArg) -> CustomType {
+    CustomType::new(ANGLE_TYPE_ID, [log_denom_arg], EXTENSION_ID, TypeBound::Eq)
+}
+
+/// Angle type with a given log-denominator (specified by the TypeArg).
+///
+/// This type is capable of representing angles that are multiples of 2π / 2^N where N is the
+/// log-denominator.
+pub(super) fn angle_type(log_denom_arg: TypeArg) -> Type {
+    Type::new_extension(angle_custom_type(log_denom_arg))
+}
+
+/// The smallest forbidden log-denominator.
+pub const LOG_DENOM_BOUND: u8 = 54;
+
+const fn is_valid_log_denom(n: u8) -> bool {
+    n < LOG_DENOM_BOUND
+}
+
+/// Type parameter for the log-denominator of an angle.
+// SAFETY: unsafe block should be ok as the value is definitely not zero.
+#[allow(clippy::assertions_on_constants)]
+pub const LOG_DENOM_TYPE_PARAM: TypeParam = TypeParam::bounded_nat(unsafe {
+    assert!(LOG_DENOM_BOUND > 0);
+    NonZeroU64::new_unchecked(LOG_DENOM_BOUND as u64)
+});
+
+/// Get the log-denominator of the specified type argument or error if the argument is invalid.
+pub(super) fn get_log_denom(arg: &TypeArg) -> Result<u8, TypeArgError> {
+    match arg {
+        TypeArg::BoundedNat { n } if is_valid_log_denom(*n as u8) => Ok(*n as u8),
+        _ => Err(TypeArgError::TypeMismatch {
+            arg: arg.clone(),
+            param: LOG_DENOM_TYPE_PARAM,
+        }),
+    }
+}
+
+pub(super) const fn type_arg(log_denom: u8) -> TypeArg {
+    TypeArg::BoundedNat {
+        n: log_denom as u64,
+    }
+}
+
+/// An angle
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
```

Comment:
I note we aren't actually deriving (or implementing) Hash here...that may be right though...